From 9a91092ba3433f6e9e2515547215abbf839382fb Mon Sep 17 00:00:00 2001
From: Pan Junzhong <junzhong.pan@spacemit.com>
Date: Mon, 20 May 2024 15:11:44 +0800
Subject: [PATCH 0694/1206] usb: dwc3-spacemit: add reset operation at standby,
 set quirk for xhci

Add reset-on-resume quirk so that xhci-plat set XHCI_RESET_ON_RESUME.

Change-Id: I3be0e06e9621d2e6a5eda3b0b332c7323522cf7c
---
 arch/riscv/boot/dts/spacemit/k1-x_MINI-PC.dts |  1 +
 arch/riscv/boot/dts/spacemit/k1-x_MUSE-N1.dts |  1 +
 arch/riscv/boot/dts/spacemit/k1-x_MUSE-Pi.dts |  1 +
 arch/riscv/boot/dts/spacemit/k1-x_deb1.dts    |  1 +
 arch/riscv/boot/dts/spacemit/k1-x_deb2.dts    |  1 +
 arch/riscv/boot/dts/spacemit/k1-x_evb.dts     |  1 +
 arch/riscv/boot/dts/spacemit/k1-x_hs450.dts   |  1 +
 arch/riscv/boot/dts/spacemit/k1-x_kx312.dts   |  1 +
 arch/riscv/boot/dts/spacemit/k1-x_mingo.dts   |  1 +
 drivers/usb/dwc3/dwc3-spacemit.c              | 17 ++++++++++++++++-
 10 files changed, 25 insertions(+), 1 deletion(-)

diff --git a/arch/riscv/boot/dts/spacemit/k1-x_MINI-PC.dts b/arch/riscv/boot/dts/spacemit/k1-x_MINI-PC.dts
index 5d921c85b07b..72648a682006 100644
--- a/arch/riscv/boot/dts/spacemit/k1-x_MINI-PC.dts
+++ b/arch/riscv/boot/dts/spacemit/k1-x_MINI-PC.dts
@@ -646,6 +646,7 @@ &gpio 123 0	/* usb3 hub en */
 
 &usbdrd3 {
 	status = "okay";
+	reset-on-resume;
 	dwc3@c0a00000 {
 		dr_mode = "host";
 		phy_type = "utmi";
diff --git a/arch/riscv/boot/dts/spacemit/k1-x_MUSE-N1.dts b/arch/riscv/boot/dts/spacemit/k1-x_MUSE-N1.dts
index 21641345482a..1c5d3d03406c 100644
--- a/arch/riscv/boot/dts/spacemit/k1-x_MUSE-N1.dts
+++ b/arch/riscv/boot/dts/spacemit/k1-x_MUSE-N1.dts
@@ -649,6 +649,7 @@ &usb3hub {
 
 &usbdrd3 {
 	status = "okay";
+	reset-on-resume;
 	dwc3@c0a00000 {
 		dr_mode = "otg";
 		phy_type = "utmi";
diff --git a/arch/riscv/boot/dts/spacemit/k1-x_MUSE-Pi.dts b/arch/riscv/boot/dts/spacemit/k1-x_MUSE-Pi.dts
index 8d1a9c365f63..cc3bf3d3bbfa 100644
--- a/arch/riscv/boot/dts/spacemit/k1-x_MUSE-Pi.dts
+++ b/arch/riscv/boot/dts/spacemit/k1-x_MUSE-Pi.dts
@@ -799,6 +799,7 @@ &usb3hub {
 
 &usbdrd3 {
 	status = "okay";
+	reset-on-resume;
 	dwc3@c0a00000 {
 		dr_mode = "host";
 		phy_type = "utmi";
diff --git a/arch/riscv/boot/dts/spacemit/k1-x_deb1.dts b/arch/riscv/boot/dts/spacemit/k1-x_deb1.dts
index 33427ba984ba..39ae6ed03a40 100644
--- a/arch/riscv/boot/dts/spacemit/k1-x_deb1.dts
+++ b/arch/riscv/boot/dts/spacemit/k1-x_deb1.dts
@@ -785,6 +785,7 @@ &gpio 123 0	/* usb3 hub en */
 
 &usbdrd3 {
 	status = "okay";
+	reset-on-resume;
 	dwc3@c0a00000 {
 		dr_mode = "host";
 		phy_type = "utmi";
diff --git a/arch/riscv/boot/dts/spacemit/k1-x_deb2.dts b/arch/riscv/boot/dts/spacemit/k1-x_deb2.dts
index acacce49cdc5..a87632a0bbb4 100644
--- a/arch/riscv/boot/dts/spacemit/k1-x_deb2.dts
+++ b/arch/riscv/boot/dts/spacemit/k1-x_deb2.dts
@@ -754,6 +754,7 @@ &gpio 123 0	/* usb3 hub en */
 
 &usbdrd3 {
 	status = "okay";
+	reset-on-resume;
 	dwc3@c0a00000 {
 		dr_mode = "host";
 		phy_type = "utmi";
diff --git a/arch/riscv/boot/dts/spacemit/k1-x_evb.dts b/arch/riscv/boot/dts/spacemit/k1-x_evb.dts
index fda758eb3b3d..ac5f50ded19c 100644
--- a/arch/riscv/boot/dts/spacemit/k1-x_evb.dts
+++ b/arch/riscv/boot/dts/spacemit/k1-x_evb.dts
@@ -713,6 +713,7 @@ &ehci1 {
 
 &usbdrd3 {
 	status = "disabled";
+	reset-on-resume;
 	dwc3@c0a00000 {
 		dr_mode = "host";
 		phy_type = "utmi";
diff --git a/arch/riscv/boot/dts/spacemit/k1-x_hs450.dts b/arch/riscv/boot/dts/spacemit/k1-x_hs450.dts
index d2b130724e9e..bb2d2073e6f6 100644
--- a/arch/riscv/boot/dts/spacemit/k1-x_hs450.dts
+++ b/arch/riscv/boot/dts/spacemit/k1-x_hs450.dts
@@ -670,6 +670,7 @@ &combphy {
 
 &usbdrd3 {
 	status = "okay";
+	reset-on-resume;
 	dwc3@c0a00000 {
 		dr_mode = "host";
 		phy_type = "utmi";
diff --git a/arch/riscv/boot/dts/spacemit/k1-x_kx312.dts b/arch/riscv/boot/dts/spacemit/k1-x_kx312.dts
index c6eade59683d..7a6f1d702ad2 100644
--- a/arch/riscv/boot/dts/spacemit/k1-x_kx312.dts
+++ b/arch/riscv/boot/dts/spacemit/k1-x_kx312.dts
@@ -642,6 +642,7 @@ &usb3hub {
 
 &usbdrd3 {
 	status = "okay";
+	reset-on-resume;
 	dwc3@c0a00000 {
 		dr_mode = "host";
 		phy_type = "utmi";
diff --git a/arch/riscv/boot/dts/spacemit/k1-x_mingo.dts b/arch/riscv/boot/dts/spacemit/k1-x_mingo.dts
index 170fa75ed9f6..99c8900e6f69 100644
--- a/arch/riscv/boot/dts/spacemit/k1-x_mingo.dts
+++ b/arch/riscv/boot/dts/spacemit/k1-x_mingo.dts
@@ -527,6 +527,7 @@ &usb3hub {
 
 &usbdrd3 {
 	status = "okay";
+	reset-on-resume;
 	dwc3@c0a00000 {
 		dr_mode = "otg";
 		phy_type = "utmi";
diff --git a/drivers/usb/dwc3/dwc3-spacemit.c b/drivers/usb/dwc3/dwc3-spacemit.c
index 6da8db13f776..b5d510b33c83 100755
--- a/drivers/usb/dwc3/dwc3-spacemit.c
+++ b/drivers/usb/dwc3/dwc3-spacemit.c
@@ -33,6 +33,7 @@ struct dwc3_spacemit {
 	struct clk		*clks[DWC3_SPACEMIT_MAX_CLOCKS];
 	int			num_clks;
 	int			suspend_clk_idx;
+	bool		reset_on_resume;
 };
 
 static int dwc3_spacemit_init(struct dwc3_spacemit *data)
@@ -109,6 +110,8 @@ static int dwc3_spacemit_probe(struct platform_device *pdev)
 		goto populate_err;
 	}
 
+	spacemit->reset_on_resume = device_property_read_bool(&pdev->dev, "reset-on-resume");
+
 	if (node) {
 		ret = of_platform_populate(node, NULL, NULL, dev);
 		if (ret) {
@@ -178,8 +181,14 @@ MODULE_DEVICE_TABLE(of, spacemit_dwc3_match);
 static int dwc3_spacemit_suspend(struct device *dev)
 {
 	struct dwc3_spacemit *spacemit = dev_get_drvdata(dev);
-	int i;
+	int i, ret;
 
+	if (spacemit->reset_on_resume){
+		ret = reset_control_assert(spacemit->resets);
+		if (ret)
+			return ret;
+		dev_info(spacemit->dev, "Will reset controller and phy on resume\n");
+	}
 	for (i = spacemit->num_clks - 1; i >= 0; i--)
 		clk_disable_unprepare(spacemit->clks[i]);
 
@@ -200,6 +209,12 @@ static int dwc3_spacemit_resume(struct device *dev)
 		}
 	}
 
+	if (spacemit->reset_on_resume){
+		dev_info(spacemit->dev, "Resetting controller and phy\n");
+		ret = reset_control_deassert(spacemit->resets);
+		if (ret)
+			return ret;
+	}
 	return 0;
 }
 
-- 
2.47.0

